File: frontend/src/shared/services/api/userService.ts
import type { UpdateUserRequest, UserResponse } from '@shared/types';
import apiClient from './client';

export const userService = {
  /**
   * Get all users (ADMIN only)
   */
  async getAllUsers(): Promise<UserResponse[]> {
    const response: any = await apiClient.get('/users');
    return response.data;
  },

  /**
   * Get user by ID
   */
  async getUserById(userId: string): Promise<UserResponse> {
    const response: any = await apiClient.get(`/users/${userId}`);
    return response.data;
  },

  /**
   * Update user
   */
  async updateUser(userId: string, data: UpdateUserRequest): Promise<UserResponse> {
    const response: any = await apiClient.put(`/users/${userId}`, data);
    return response.data;
  },

  /**
   * Delete user (ADMIN only)
   */
  async deleteUser(userId: string): Promise<void> {
    await apiClient.delete(`/users/${userId}`);
  },
};

File: backend/src/main/java/com/example/backendservice/features/user/service/UserService.java
package com.example.backendservice.features.user.service;

import com.example.backendservice.features.user.dto.*;
import com.example.backendservice.features.user.entity.RoleType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface UserService {

    // User CRUD
    UserResponse createUser(CreateUserRequest request);

    UserResponse getUserById(UUID userId);

    UserResponse getUserByEmail(String email);

    UserResponse updateUser(UUID userId, UpdateUserRequest request);

    void deleteUser(UUID userId);

    // User listing
    Page<UserResponse> getAllUsers(Pageable pageable);

    Page<UserResponse> getUsersByRole(RoleType role, Pageable pageable);

    // Citizen profile
    CitizenProfileResponse getCitizenProfile(UUID userId);

    CitizenProfileResponse updateCitizenProfile(UUID userId, UpdateCitizenProfileRequest request);

    // Collector profile
    CollectorProfileResponse getCollectorProfile(UUID userId);

    CollectorProfileResponse updateCollectorProfile(UUID userId, UpdateCollectorProfileRequest request);

    CollectorProfileResponse updateCollectorLocation(UUID userId, Double lat, Double lng);

    CollectorProfileResponse setCollectorAvailability(UUID userId, Boolean isAvailable);

    // Password management
    void changePassword(UUID userId, String oldPassword, String newPassword);

    void resetPassword(String email);
}
